Stop gdbserver after debug session

ArsenInstigate opened the pull request:
Added disconnectRequest function which stops gdbserver after debug session

jonahgraham commented on src/GDBTargetDebugSession.ts:
```diff
@@ -341,4 +341,21 @@ export class GDBTargetDebugSession extends GDBDebugSession {
+    ): Promise<void> {
+        try {
+            await this.gdb.sendGDBExit();
+            await this.gdbserver?.kill();
```
Adding this looks fine to me. Is there any issue with how gracefully we shutdown openocd? 

This `kill()` does a `SIGTERM` which can be a graceful exit?

I have tested this on Linux with `gdbserver` and it works for me. I don't have an openocd device to hand.

jonahgraham replied:
Another question - do we need to wait for the server to exit. e.g. if the user presses restart VSCode will send `disconnect` request and then once disconnect response is sent VSCode will start again. But if the gdbserver does not exit quickly enough we will have the same problem?

Should we listen to the [`exit`](https://nodejs.org/api/child_process.html#event-exit) event of the process with a timeout.

Keep in mind that if we add that functionality that VSCode has a [2 second timeout](https://github.com/Microsoft/vscode/blob/e9229a18805fdb9baf7356ddc31bd216da5842ef/src/vs/workbench/contrib/debug/browser/rawDebugSession.ts#L572) to get a response to a disconnect request.

ArsenInstigate commented:
Hi @jonahgraham 

I added the following futures, to my pull request

1. Moved stop gdb server to separate function.
2. Added timeout after killing gdbserver.
3. Added test in integration-tests for stop gdb server.

On my board, it works. So what should be my next step?

jonahgraham commented:
> So what should be my next step?

Please sign the [ECA](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/pull/258#issuecomment-1450198496)

This change looks fine to me, I have pushed a couple of extra commits to fix formatting and make it possible to disable the new functionality as it seems to interfere with Renesas' use case in #258

jonahgraham commented:
The PR run failed with this change on Windows. It looks like a probable flaky test so I have retriggered it.

jonahgraham commented:
Thanks @ArsenInstigate!

omarArm commented:
Hi All,

So I am trying to create a new "remote target" test which is supposed to be very similar to the test that was added in this PR. I have a small question though, how do the integration tests handle "remote target" tests exactly? 

I was under the impression that I would need to provide some sort of gdbserver  and a target.port to connect to, so having something like this:
```
await dc.launchRequest(
                fillDefaults(this.currentTest, {
                    program: join(testProgramsDir, 'empty'),
                    target: {
                        port: <some number>,
                        type: 'remote',
                        serverParameters: ['some server'],
                    }
                } as unknown as TargetLaunchRequestArguments)
        ); 
```

However, what I see in this PR is the following 
```
fillDefaults(this.test, {
                program: emptyProgram,
                target: {
                    type: 'remote',
                    serverParameters: [':0', emptyProgram],
                } as TargetLaunchArguments,
            } as TargetLaunchRequestArguments);
```

In which, no port number for the client is provided, just the port number for the server as part of the server parameters. How are the tests handling this client-server connection exactly? and what am I missing?

Thank in advance for your support and help

jonahgraham commented:
@omarArm My guess is your missing here is that the default remote target will auto-start gdbserver when doing a "launch" type of connection as is done in this PR (see implementation of [hitBreakpoint](https://github.com/microsoft/vscode-debugadapter-node/blob/950789a5313424bee7a126982f49b6ab19781eea/testSupport/src/debugClient.ts#L478) where it calls [launch](https://github.com/microsoft/vscode-debugadapter-node/blob/950789a5313424bee7a126982f49b6ab19781eea/testSupport/src/debugClient.ts#L342) which calls launchRequest)

So in this case, since it is a launch, the `server` parameter is important:

https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/1a5a38d39e4ad3e6f93214c584c3b80263495034/src/types/session.ts#L131-L133

The `serverPortRegExp` is used to *grep* the output of the server launching to extract the port to connect to:

https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/1a5a38d39e4ad3e6f93214c584c3b80263495034/src/types/session.ts#L137-L141

---

[remote-target tests](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/135f8621edaf939befcac528ac8c29479e83a27d/package.json#L30) set [isRemoteTest](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/590fea58cfb2ee4984d766cd1b2140738d3ff110/src/integration-tests/utils.ts#L218) which controls which adapter is launched:

https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/590fea58cfb2ee4984d766cd1b2140738d3ff110/src/integration-tests/utils.ts#L314-L319

---

Does that answer your open questions?

omarArm commented:
Hi Jonah, 
thank you so much for your answer. Yes! I finally get it. Thanks again
